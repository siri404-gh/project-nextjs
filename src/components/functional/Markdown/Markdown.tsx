import React, { FunctionComponent } from 'react';
import ReactMarkdown, { ReactMarkdownProps as MarkdownProps } from 'react-markdown';
import {
  Typography,
  Link,
  makeStyles,
} from '@material-ui/core';

import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { okaidia as theme } from 'react-syntax-highlighter/dist/cjs/styles/prism';

const useStyles = makeStyles(theme => ({
  li: {
    // eslint-disable-next-line no-magic-numbers
    marginTop: theme.spacing(1),
  },
  p: {
    fontWeight: theme.typography.fontWeightLight,
    textAlign: 'justify',
  },
}));

const renderers = {
  code: ({ language, value }) => {

    return <SyntaxHighlighter style={theme} language={language} showLineNumbers={false} customStyle={{ fontSize: 12 }}>{value}</SyntaxHighlighter>
  },
  link: ({ href, children }) => {
    return <Link href={href}>{children}</Link>
  },
  heading: ({ children, level }) => {

    const getComponent = level => {
      switch (level) {
        case 1:
          return 'h1';
        case 2:
          return 'h2';
        case 3:
          return 'h3';
        case 4:
          return 'span';
      }
    };

    const getVariant = level => {
      switch (level) {
        case 1:
          return 'h5';
        case 2:
          return 'h6';
        case 3:
          return 'subtitle1';
        case 4 :
          return 'caption';
      }
    };

    return <Typography component={getComponent(level)} variant={getVariant(level)} gutterBottom>{children}</Typography>
  },
  paragraph: ({ children }) => {
    const classes = useStyles();
    return <Typography classes={{ body2: classes.p }} component='p' variant='body2' paragraph>{children}</Typography>
  },
  span: ({ children }) => {
    const classes = useStyles();
    return <Typography classes={{ body2: classes.p }} component='span' variant='body2' paragraph>{children}</Typography>;
  },
  listItem: ({ children }) => {
    const classes = useStyles();
    return (
      <li className={classes.li}>
        <Typography component='span'>{children}</Typography>;
      </li>
    );
  },
};

const Markdown: FunctionComponent<MarkdownProps> = ({ children }) => {
  return <ReactMarkdown renderers={renderers} allowDangerousHtml>{children}</ReactMarkdown>
};

export default Markdown;
