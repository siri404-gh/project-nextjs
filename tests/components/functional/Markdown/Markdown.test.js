import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';

import Markdown from '@/components/functional/Markdown/Markdown';

describe('Markdown', () => {
  let container;

  beforeEach(() => {
    container = document.createElement('section');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('Should render without props', () => {
    act(() => {
      render(<Markdown />, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render a', () => {
    act(() => {
      render(<Markdown>
        {"[#](#)"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render h1', () => {
    act(() => {
      render(<Markdown>
        {"# h1"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render h2', () => {
    act(() => {
      render(<Markdown>
        {"## h2"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render h3', () => {
    act(() => {
      render(<Markdown>
        {"### h3"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render h4', () => {
    act(() => {
      render(<Markdown>
        {"#### h4"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render p', () => {
    act(() => {
      render(<Markdown>
        {"p"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render span', () => {
    act(() => {
      render(<Markdown>
        {"<span>span</span>"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render li', () => {
    act(() => {
      render(<Markdown>
        {" - li"}
      </Markdown>, container);
    });
    expect(container).toMatchSnapshot();
  });
});
