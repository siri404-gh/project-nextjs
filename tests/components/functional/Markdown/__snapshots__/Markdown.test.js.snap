// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown Should render a 1`] = `
<section>
  <p
    class="MuiTypography-root MuiTypography-body2 makeStyles-p-2 MuiTypography-paragraph"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
      href="#"
    >
      #
    </a>
  </p>
</section>
`;

exports[`Markdown Should render h1 1`] = `
<section>
  <h1
    class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
  >
    h1
  </h1>
</section>
`;

exports[`Markdown Should render h2 1`] = `
<section>
  <h2
    class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom"
  >
    h2
  </h2>
</section>
`;

exports[`Markdown Should render h3 1`] = `
<section>
  <h3
    class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom"
  >
    h3
  </h3>
</section>
`;

exports[`Markdown Should render h4 1`] = `
<section>
  <span
    class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom"
  >
    h4
  </span>
</section>
`;

exports[`Markdown Should render li 1`] = `
<section>
  <ul>
    <li
      class="makeStyles-li-7"
    >
      <span
        class="MuiTypography-root MuiTypography-body1"
      >
        li
      </span>
    </li>
  </ul>
</section>
`;

exports[`Markdown Should render p 1`] = `
<section>
  <p
    class="MuiTypography-root MuiTypography-body2 makeStyles-p-4 MuiTypography-paragraph"
  >
    p
  </p>
</section>
`;

exports[`Markdown Should render span 1`] = `
<section>
  <p
    class="MuiTypography-root MuiTypography-body2 makeStyles-p-6 MuiTypography-paragraph"
  >
    <span>
      span
    </span>
  </p>
</section>
`;

exports[`Markdown Should render without props 1`] = `<section />`;
